### PR TITLE
Rename --out-dir to --artifact-dir

### PR DESCRIPTION
**Pull request by valadaptive**
Progress towards unblocking #6790. Renames the experimental `--out-dir` argument to `--artifact-dir`, both to reflect that it's where the final build *artifacts* will be copied to, and to avoid confusion with the `OUT_DIR` environment variable which serves an entirely different purpose.

For transition purposes, `--out-dir` argument and `out-dir` config key will still work with a deprecation message encouraging the use of the new arg and config key.

### Rationale

A lot of people seem to be confused by the naming of the `--out-dir` argument, and are misled into thinking it serves the same purpose as the `OUT_DIR` environment variable:

> [However, it doesn't seem that OUT_DIR environment variable is set to the value of --out-dir when build.rs is executed.](https://github.com/rust-lang/cargo/issues/6790#issuecomment-574754790)

> [I understand that the worry is that there could be confusion between --out-dir for cargo and the environment variable OUT_DIR for build.rs, but doesn't it mean exactly the same in both cases?](https://github.com/rust-lang/cargo/issues/6790#issuecomment-2091259652)

> [--out-dir: Things will be built into $PWD/target as normal, but copies some of the artifacts into the directory specified by out-dir (not a profile specific subdirectory). Unstable flag, added in March 2018. cargo build --out-dir #5203 Ability to specify output artifact name #4875. **Mimicks the behavior of OUT_DIR.**](https://github.com/rust-lang/cargo/issues/6100#issue-364115234)

> [I recently had a couple of people express an interest in --out-dir being stabilized and from my initial digging it seems like what they may actually want is to switch to OUT_DIR, which is already stable.](https://github.com/rust-lang/cargo/issues/6100#issuecomment-820783316)

### CODE REVIEW

**Comment by workingjubilee** [comment]
Grim news. Perhaps both flags must be briefly supported to deal with the bootstrap problem.

**Comment by valadaptive** [comment]
Never mind; it seems I was a bit zealous in replacing `--out-dir` and accidentally search-and-replaced it in some rustc invocations. Might still need to do some fixups but we're looking a lot better now.

**Comment by valadaptive** [comment]
This should be ready for review now. Not sure if this needs to go through any sort of design process since it's renaming a nightly-only API. There does seem to be some consensus in #6790 that `--artifact-dir` is a better name.

**Comment by valadaptive** [comment]
@epage Just wanted to make sure this hasn't fallen off your radar--it's ready for review now.

**Comment by epage** [comment]
I have about a month backlog of github notifications that I still need to respond to.

**Comment by epage** [comment]
Could you update the commits for how you'd want them presented for review and merging?

File: src/bin/cargo/commands/build.rs [hunk 1]
```diff
@@ -50,14 +50,45 @@ pub fn exec(gctx: &mut GlobalContext, args: &ArgMatches) -> CliResult {
...
-    if let Some(out_dir) = args.value_of_path("out-dir", gctx) {
-        compile_opts.build_config.export_dir = Some(out_dir);
-    } else if let Some(out_dir) = gctx.build_config()?.out_dir.as_ref() {
-        let out_dir = out_dir.resolve_path(gctx);
-        compile_opts.build_config.export_dir = Some(out_dir);
+    // If the user specifies `--artifact-dir`, use that
+    let mut artifact_dir = args.value_of_path("artifact-dir", gctx);
+
+    // `--out-dir` is deprecated, but still supported for now
+    if artifact_dir.is_none() {
```
Comment (epage) [inline src/bin/cargo/commands/build.rs:57]: Shouldn't the `if/else-if` approach still work?  Keeping the structure would make it easier to read and this embeds higher level context into the code structure (makes it clear to readers that they are mutually exclusive)

Comment (epage) [reply src/bin/cargo/commands/build.rs:57]: You could even combine the approaches and do
```rust
compile_opts.build_config.export_dir = if let Some(artifact_dir) = ...
};
```

Comment (valadaptive) [reply src/bin/cargo/commands/build.rs]: I've changed it back to the old code style.

The if-expression style may not be equivalent to the existing code--right now `compile_opts.build_config.export_dir` is preserved if none of the command-line flags or config flags are present, whereas the if-expression would unconditionally reassign (in the final `else`, we *could* just assign it to its existing value, but that feels too "clever"). I don't *think* `compile_opts.build_config.export_dir` is set by any other code, but I'd rather not risk it.

**Comment by valadaptive** [comment]
I've updated the flag-checking code and squashed most of the commits.

Should the deprecation messages provide some sort of warning that the old options will eventually be removed, or are they fine currently?

File: tests/testsuite/artifact_dir.rs [hunk 2]
```diff
@@ -304,6 +304,55 @@ For more information, try '--help'.
         .run();
 }
 
+#[cargo_test]
+fn deprecated_out_dir() {
```
Comment (epage) [inline tests/testsuite/artifact_dir.rs:308]: nit: not important enough for this change but in the future, I generally recommend adding the tests as  the first commit (with them passing).  Then as you change behavior, that is reflected in how the tests are changed to ensure they still pass.  This makes it easier for reviewers to see how behavior changed, we can then show those diffs in posts to the community for them to understand changes, and it helps "test the test" (I've made changes where I thought I was testing the right condition and wasn't).

**Pull request merged**